onnxruntime/core/providers/rocm/roctracer_manager.cc

```diff
@@ -0,0 +1,465 @@
+      auto const& launch_args = call_record.api_data_.args;
+      name = roctracer_op_string(call_record.domain_, call_record.cid_, 0);
+
+      args = {
+          {"stream", call_record.cid_ == HIP_API_ID_hipMemset
+                         ? "0"
+                         : PointerToHexString((void*)launch_args.hipMemsetAsync.stream)},
+          {"dst", PointerToHexString(launch_args.hipMemset.dst)},
+          {"size", std::to_string(launch_args.hipMemset.sizeBytes)},
+          {"value", std::to_string(launch_args.hipMemset.value)}};
+      break;
+    }
+
+    case HIP_API_ID_hipMemcpy:
+    case HIP_API_ID_hipMemcpyAsync:
+    case HIP_API_ID_hipMemcpyWithStream: {
+      auto const& launch_args = call_record.api_data_.args;
+      name = roctracer_op_string(call_record.domain_, call_record.cid_, 0);
+
+      args = {
+          {"stream", call_record.cid_ == HIP_API_ID_hipMemcpy
+                         ? "0"
+                         : PointerToHexString((void*)launch_args.hipMemcpyAsync.stream)},
+          {"src", PointerToHexString(launch_args.hipMemcpy.src)},
+          {"dst", PointerToHexString(launch_args.hipMemcpy.dst)},
+          {"kind", MemcpyKindToString(launch_args.hipMemcpy.kind)}};
+      break;
+    }
+
+    case HIP_API_ID_hipMemcpy2D:
+    case HIP_API_ID_hipMemcpy2DAsync: {
+      auto const& launch_args = call_record.api_data_.args;
+      name = roctracer_op_string(call_record.domain_, call_record.cid_, 0);
+
+      args = {
+          {"stream", call_record.cid_ == HIP_API_ID_hipMemcpy2D
+                         ? "0"
+                         : PointerToHexString((void*)launch_args.hipMemcpy2DAsync.stream)},
+          {"src", PointerToHexString(launch_args.hipMemcpy2D.src)},
+          {"dst", PointerToHexString(launch_args.hipMemcpy2D.dst)},
+          {"spitch", std::to_string(launch_args.hipMemcpy2D.spitch)},
+          {"dpitch", std::to_string(launch_args.hipMemcpy2D.dpitch)},
+          {"width", std::to_string(launch_args.hipMemcpy2D.width)},
+          {"height", std::to_string(launch_args.hipMemcpy2D.height)},
+          {"kind", MemcpyKindToString(launch_args.hipMemcpy2D.kind)}};
+      break;
+    }
+
+    case HIP_API_ID_hipExtModuleLaunchKernel: {
+      auto const& launch_args = call_record.api_data_.args.hipExtModuleLaunchKernel;
+      name = demangle(hipKernelNameRef(launch_args.f));
+
+      args = {
+          {"stream", PointerToHexString((void*)launch_args.hStream)},
+          {"grid_x", std::to_string(launch_args.globalWorkSizeX)},
+          {"grid_y", std::to_string(launch_args.globalWorkSizeY)},
+          {"grid_z", std::to_string(launch_args.globalWorkSizeZ)},
+          {"block_x", std::to_string(launch_args.localWorkSizeX)},
+          {"block_y", std::to_string(launch_args.localWorkSizeY)},
+          {"block_z", std::to_string(launch_args.localWorkSizeZ)},
+      };
+      break;
+    }
+
+    default:
+      return false;
+  }
+
+  new (&event) EventRecord{
```

Comment:
Can you elaborate why a placement new is needed here?

Reply:
The alternative is to construct a temporary `EventRecord` object and assign it to `event`. The placement new constructs the `EventRecord` in-place, eliminating the assignment.